File: web/src/sdk/driver-manifest.ts
export enum Protocol {
  GPIO = "GPIO",
  Serial = "Serial",
  USB = "USB",
  I2C = "I2C",
  SPI = "SPI",
  CAN = "CAN",
  JTAG = "JTAG",
  Bluetooth = "Bluetooth",
  Network = "Network",
}

export type SafetyLevel = "safe" | "unsafe" | "critical";

export interface HalTraitMethod {
  name: string;
  signature: string;
}

export interface HalTraitDefinition {
  name: string;
  methods: HalTraitMethod[];
  safety_level: SafetyLevel;
}

export interface DriverManifest {
  id: string;
  name: string;
  version: string;
  description: string;
  author: string;
  protocols: Protocol[];
  hal_traits: HalTraitDefinition[];
  entrypoint: string;
}

export interface ValidationResult {
  valid: boolean;
  errors: string[];
}

export function validateDriverManifest(manifest: unknown): ValidationResult {
  const errors: string[] = [];

  if (!manifest || typeof manifest !== "object") {
    return { valid: false, errors: ["manifest must be an object"] };
  }

  const m = manifest as Partial<DriverManifest>;

  if (typeof m.id !== "string" || m.id.trim().length === 0) {
    errors.push("id is required and must be a non-empty string");
  }
  if (typeof m.name !== "string" || m.name.trim().length === 0) {
    errors.push("name is required and must be a non-empty string");
  }
  if (typeof m.version !== "string" || m.version.trim().length === 0) {
    errors.push("version is required and must be a non-empty string");
  }
  if (typeof m.description !== "string") {
    errors.push("description is required and must be a string");
  }
  if (typeof m.author !== "string" || m.author.trim().length === 0) {
    errors.push("author is required and must be a non-empty string");
  }
  if (!Array.isArray(m.protocols)) {
    errors.push("protocols is required and must be an array");
  }
  if (!Array.isArray(m.hal_traits)) {
    errors.push("hal_traits is required and must be an array");
  } else {
    m.hal_traits.forEach((t, i) => {
      if (!t || typeof t.name !== "string") {
        errors.push(`hal_traits[${i}].name is required`);
      }
      if (!t || !Array.isArray(t.methods)) {
        errors.push(`hal_traits[${i}].methods must be an array`);
      }
      if (!t || typeof t.safety_level !== "string") {
        errors.push(`hal_traits[${i}].safety_level is required`);
      }
    });
  }
  if (typeof m.entrypoint !== "string" || m.entrypoint.trim().length === 0) {
    errors.push("entrypoint is required and must be a non-empty string");
  }

  return { valid: errors.length === 0, errors };
}

export function createDefaultDriverManifest(id: string, name: string, author: string): DriverManifest {
  return {
    id,
    name,
    version: "0.1.0",
    description: `${name} driver`,
    author,
    protocols: [Protocol.USB],
    hal_traits: [],
    entrypoint: "libdriver.so",
  };
}

File: crates/hal-core/src/driver_manifest.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Protocol {
    GPIO,
    Serial,
    USB,
    I2C,
    SPI,
    CAN,
    JTAG,
    Bluetooth,
    Network,
}

impl Protocol {
    pub fn as_str(&self) -> &'static str {
        match self {
            Protocol::GPIO => "GPIO",
            Protocol::Serial => "Serial",
            Protocol::USB => "USB",
            Protocol::I2C => "I2C",
            Protocol::SPI => "SPI",
            Protocol::CAN => "CAN",
            Protocol::JTAG => "JTAG",
            Protocol::Bluetooth => "Bluetooth",
            Protocol::Network => "Network",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum SafetyLevel {
    Safe,
    Unsafe,
    Critical,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HalTraitMethod {
    pub name: String,
    pub signature: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HalTraitDefinition {
    pub name: String,
    pub methods: Vec<HalTraitMethod>,
    pub safety_level: SafetyLevel,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DriverManifest {
    pub id: String,
    pub name: String,
    pub version: String,
    pub description: String,
    pub author: String,
    pub protocols: Vec<Protocol>,
    pub hal_traits: Vec<HalTraitDefinition>,
    pub entrypoint: String,
}

impl DriverManifest {
    pub fn from_json(s: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(s)
    }

    pub fn to_json(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string_pretty(self)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn serialize_roundtrip() {
        let manifest = DriverManifest {
            id: "drv-1".into(),
            name: "Example Driver".into(),
            version: "0.1.0".into(),
            description: "An example driver".into(),
            author: "prometheus".into(),
            protocols: vec![Protocol::USB, Protocol::Serial],
            hal_traits: vec![HalTraitDefinition {
                name: "Probeable".into(),
                methods: vec![HalTraitMethod {
                    name: "probe".into(),
                    signature: "fn probe(&self, target: &str)".into(),
                }],
                safety_level: SafetyLevel::Safe,
            }],
            entrypoint: "libdriver.so".into(),
        };
        let json = manifest.to_json().unwrap();
        let back = DriverManifest::from_json(&json).unwrap();
        assert_eq!(back.id, "drv-1");
        assert_eq!(back.protocols.len(), 2);
    }
}
